### PR TITLE
passthrough_ll update

### PR DESCRIPTION
**Pull request by szmi**
This fixes a couple of bugs in the passthrough_ll example filesystem and adds support for lots of missing operations.  Options are added to configure a base path and to control caching.

### CODE REVIEW

File: example/passthrough_ll.c [hunk 1]
```diff
@@ -87,6 +88,9 @@ static const struct fuse_opt lo_opts[] = {
 	  offsetof(struct lo_data, writeback), 1 },
 	{ "no_writeback",
 	  offsetof(struct lo_data, writeback), 0 },
+	{ "source=%s",
```
Comment (Nikratio) [inline example/passthrough_ll.c]: I'd much rather see this as a parameter (like the mountpoint). Any reason you made it a mount option?

**Comment by Nikratio** [comment]
Thanks! This is great, I really like having a more feature-rich low-level example. My only concerns are the  two minor issues mentioned above.

**Comment by Nikratio** [comment]
*ping* Would you be able to add the documentation update, so this can get merged? Should be just a few lines if I understand the issue correctly.

**Comment by Nikratio** [comment]
*ping* Are you no longer interested in working on this?

The only thing that prevents a merge is the missing testcase...

**Comment by szmi** [comment]
I'm definitely interested in this getting merged.  It's part of a bigger project that we are working on internally and will hopefully go public in some weeks. Getting it upstream is obviously very important.

OTOH, I've a lot on my plate and python is not the language that I'm highly skilled in, so I'd definitely accept some help there.

**Comment by Nikratio** [comment]
Got you. Will add a test case later today, if I can find out how to extend the push request (there was some way to do that...)

**Comment by Nikratio** [comment]
Ok, I've extended the testcases. There is one new failure when running with active writeback. It comes from the `test_syscalls` test which is in C :-).

```
50 [open_acc(O_RDONLY) mode: 0400 message: 'Success'] OK
 51 [open_acc(O_WRONLY) mode: 0200 message: 'Success'] do_test_open_acc() - open: Permission denied
 52 [open_acc(O_RDONLY | O_TRUNC) mode: 0400 message: 'Permission denied'] OK
```

Unfortunately I do not have permission to extend your pull request with the test changes. Could you give me permissions? Instructions at 
https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

**Comment by Nikratio** [comment]
For now, I have pushed the updated branch to https://github.com/libfuse/libfuse/commits/miklos-pull-request

**Comment by szmi** [comment]
Thank's for helping.

Ticked the box to give permissions.

The failed testcase is expected.  This mode only works if the server has elevated privileges (CAP_DAC_OVERRIDE) because in writeback mode the kernel may also issue read requests (even if file was opened for write-only) and hence the passthrough fs has to open the file for reading as well as for writing, which fails if the file is only openable for write and the server lacks privileges.

I think the right solution to that would be to disable writeback cache for the inode where read permission is denied (an unlikely situation).  This is not currently possible, as writeback mode is per-superblock not per-inode, but that's fixable in the fuse kernel module and in API's.

**Comment by Nikratio** [comment]
Right. I don't suppose you have a patch ready for implementing per-inode cache settings? :-)

What do you think is the best way to handle the current situation? I think we can either drop this particular test completely, or add an additional command line parameter to `test_syscalls` that causes it to skip the test only if writeback caching is enabled.

**Comment by szmi** [comment]
Adding "-51" to the test command line will disable this test.  If you only add it for the "-owriteback" case, all the better.

**Comment by Nikratio** [comment]
done, let's see

**Comment by Nikratio** [comment]
Are those numbers dynamic? Test 51 is now skipped, but now test 52 is the failing one?!

**Comment by szmi** [comment]
Umm, a seekdir test was added by commit 5f125c5e6be2 "Fix readdir() bug when a non-zero offset is specified in filler (#269)".  The problem is that it was added in the middle, thus causing all the test numbers after it to shift by one.

**Comment by Nikratio** [comment]
I don't follow. The test passes locally (with 51 excluded), but did not pass on Travis. I haven't rebased anything, it seems that the numbers are simply different between the test runs.

**Comment by szmi** [comment]
The travis run has a seekdir test at position12 while the local run doesn't have it.  So I guess travis is doing a merge with HEAD before running the tests.

**Comment by Nikratio** [comment]
Explicitly rebased and updated the test number now. Let's see...

**Comment by Nikratio** [comment]
Looks good now! Could you still document the  that . and .. should not get an increased lookup count? As far as I can tell, this isn't documented anywhere yet the example refers to this as a requirement.

**Comment by Nikratio** [comment]
*ping*

Any chance you could write a few words so we can merge this? :-)

**Comment by szmi** [comment]
It seems to be documented in fuse_lowlevel.h:

	 * In contrast to readdir() (which does not affect the lookup counts),
	 * the lookup count of every entry returned by readdirplus(), except "."
	 * and "..", is incremented by one.

The sentence is structured a bit awkwardly, maybe it should be
 
	 * In contrast to readdir() the lookup count of every entry returned by readdirplus() is incremented by one, with the exception of "." and ".."

**Comment by Nikratio** [comment]
Right you are. Weird, I was sure I carefully looked at exactly that section of the file...

Thanks for you patience! Merging now.

**Pull request merged**